stepsfunction/handler.py

```diff
@@ -163,19 +178,186 @@ def ocr_page(event, context):
+        # update db
+        pdf_upload.update_with_log(
+            actions=[
+                PDFUpload.status.set("ocred"),
+                PDFUpload.updatedAt.set(datetime.now())
+            ])
+        send_task_ocr_activity(jid)
+
+
+def get_textract_data(bucket, key):
+    """Using AWS textract."""
+    LOG.info(f'Loading get_textract_data bucket:{bucket}, key:{key}')
+    # convert to jpg
+    jpg_file = extract_jpg_from_pdf(bucket, key)
+    # Call Amazon Textract
+    tx = boto3.client(
+        service_name='textract',
+        region_name='us-east-1',
+        endpoint_url='https://textract.us-east-1.amazonaws.com',
+    )
+    response = tx.detect_document_text(
+        Document={
+            'Bytes': jpg_file
+        })
+    detected_text = ''
+
+    # Print detected text
+    for item in response['Blocks']:
+        if item['BlockType'] == 'LINE':
+            detected_text += item['Text'] + '\n'
+    return detected_text
+
+
+def write_textract_to_s3(textract_data, bucket, key):
+    """Save detected text to S3."""
+    LOG.info(f'Loading write_textract_to_s3 bucket:{bucket}, key:{key}')
+    generate_path = os.path.splitext(key)[0] + '.txt'
+    s3client = boto3.client('s3')
+    s3client.put_object(Body=textract_data, Bucket=bucket, Key=generate_path)
+    LOG.info(f'generateFilePath: {generate_path}')
+
+
+def update_db_after_ocr(pdf_upload, content, name, jid):
+    """Update PDFUpload after finishing orc."""
+    page = SinglePage(
+        page_id=name[:4],
+        content=content
+    )
+    pages = pdf_upload.pages
+    if pages:
+        pages.append(page)
+    else:
+        pages = [page]
+    LOG.info(f'loading update_db_after_ocr pages:{pages}')
+    pdf_upload.update_with_log(
+        actions=[
+            PDFUpload.pages.set(pages),
+            PDFUpload.updatedAt.set(datetime.now())
+        ])
+
+
+def check_orc_done(pdf_upload, jid):
+    """Get db & check done."""
+    LOG.info(f'loading check_orc_done')
+    num_pages = pdf_upload.num_pages
+    pages = pdf_upload.pages
+    pages_len = 0 if not pages else len(pages)
+    LOG.info(f'check_orc_done: num_pages:{num_pages}, pages_len={pages_len}')
+    if num_pages == pages_len:
+        return True
+    else:
+        return False
+
+
+def send_task_ocr_activity(worker_name):
+    """After ocr's done, send task success to WaitOcr activity."""
+    try:
+        activity_arn = os.environ.get('ACTIVITY_OCR_DONE_ARN', "")
+        sf = boto3.client('stepfunctions')
+        response = sf.get_activity_task(
+            activityArn=activity_arn,
+            workerName=worker_name
+        )
+        LOG.info(f'activityArn={activity_arn}, sf_activity={response}')
+        sf_input = dumps(response['input'])
+        res = sf.send_task_success(
+            taskToken=response["taskToken"],
+            output=sf_input
+        )
         LOG.info(f'sf.send_task_success={res}')
+    except ClientError as e:
+        LOG.error(e)
```

Comment:
If we only log it, we "swallow" the error. Shouldn't we propogate it or handle it so the Lambda fails? Ideally, this would trigger a failure path in the state machine like I demonstrated today. Not sure how the failure-handling paths would work yet, but if you have ideas, let's work on a GDoc.